muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java

```diff
@@ -205,6 +215,38 @@ public WorkspaceUserEntity updateIdentifier(WorkspaceUserEntity workspaceUserEnt
   }
 
   public WorkspaceUserEntity updateActive(WorkspaceUserEntity workspaceUserEntity, Boolean active) {
+    if (Boolean.FALSE.equals(active)) {
+      // #6620: Remove past workspaces from the list of student's last workspaces
+      UserEntity userEntity = userEntityController.findUserEntityByDataSourceAndIdentifier(workspaceUserEntity.getUserSchoolDataIdentifier().getDataSource(), workspaceUserEntity.getUserSchoolDataIdentifier().getIdentifier());
+      UserEntityProperty lastWorkspaces = userEntityController.getUserEntityPropertyByKey(userEntity, "last-workspaces");
+      
+      if (lastWorkspaces != null) {
+        if (StringUtils.isEmpty(lastWorkspaces.getValue())) {
+          ObjectMapper objectMapper = new ObjectMapper();
+          List<LastWorkspace> lastWorkspaceList = null;
+          String lastWorkspaceJson = lastWorkspaces.getValue();
+          
+          try {
+           lastWorkspaceList = objectMapper.readValue(lastWorkspaceJson, new TypeReference<ArrayList<LastWorkspace>>() {});
+          }
+          catch (JsonProcessingException e) {
+            logger.log(Level.WARNING, String.format("Parsing last workspaces of user %d failed: %s", userEntity.getId(), e.getMessage()));
+          }
+          
+          if(lastWorkspaceList != null) {
+            Long workspaceEntityId = workspaceUserEntity.getWorkspaceEntity().getId();
+            lastWorkspaceList.removeIf(item -> workspaceEntityId.equals(item.getWorkspaceId()));
+          }
+          
+          try {
```

Comment:
Might as well compare list size before and after modification; if it's the same, no need to update the property as nothing has changed.

Reply:
Also, if lastWorkspaceList was null (the previous if block), is there any need to save?

These all could probably be combined under one try-catch too as they are depending on each previous step in the sequence.